pkg/kubelet/cm/memorymanager/policy_static.go

```diff
@@ -157,6 +157,13 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 		bestHint = extendedHint
 	}
 
+	// the best hint might violate the NUMA allocation rule on which
+	// NUMA node cannot have both single and cross NUMA node allocations
+	// https://kubernetes.io/blog/2021/08/11/kubernetes-1-22-feature-memory-manager-moves-to-beta/#single-vs-cross-numa-node-allocation
+	if isAffinityViolatesNUMAAllocations(machineState, bestHint.NUMANodeAffinity) {
```

Comment:
I wonder if we can merge this check in the `getDefaultHint` or `extendTopologyHint` helpers

Reply:
`extendTopoloyHint` is not a good option, because it depends on `isAffinitySatisfyRequest` value, which is exactly where the bug is hiding.
In the bug flow the `if !isAffinitySatisfyRequest(machineState, bestHint.NUMANodeAffinity, requestedResources) {` line evaluate to `false`, we skip the `if` block and of course we also skip `extendTopoloyHint` helper.

`getDefaultHint` is not good enough either since it depends on `if hint.NUMANodeAffinity == nil {` while in the bug flow `hint.NUMANodeAffinity != nil`

A valid option that we might consider is always calling `extendTopologyHint` but I'm not sure about the implication of such change.